# Cálculos de speedups y eficiencia

In [1]:
# libraries
import pandas as pd
import pandas as pd
from IPython.display import display, HTML

In [2]:
# display of result tables
def display_table(name, dataframe):
    title_html = f"<h2>{name}</h2>"
    display(HTML(title_html))
    display(dataframe)

## Cálculo de Speedups
### Bruteforce y ambos acercamientos

In [3]:
# format: (brute force time, approach 1 time, approach 2 time)
execution_times = {
    "42": (0.00, 0.00, 0.00),
    "123456": (0.12, 0.16, 0.33),
    "987654": (0.95, 1.36, 2.82),
    "9876547": (19.19, 31.97, 14.55),
    "98965471": (102.24, 167.76, 377.35)
}

speedups = {}

# calculate speedups
for key, (naive, approach1, approach2) in execution_times.items():
    if naive != 0.0:
        speedup_approach1 = naive / approach1
        speedup_approach2 = naive / approach2

        # store the speedups
        speedups[key] = (speedup_approach1, speedup_approach2)

# dataframe for speedups
speedup_df = pd.DataFrame(speedups, index=["Speedup Approach 1", "Speedup Approach 2"]).T

display_table(name="Speedups Table", dataframe=speedup_df)

Speedup Approach 1  Speedup Approach 2
123456              0.750000            0.363636
987654              0.698529            0.336879
9876547             0.600250            1.318900
98965471            0.609442            0.270942

## Cálculo de Speedups
### Naive y ambos acercamientos

In [4]:
# format: (naive time, approach 1 time, approach 2 time)
execution_times = {
    "42": (0.00, 0.00, 0.00),
    "123456": (0.12, 0.16, 0.33),
    "987654": (0.91, 1.36, 2.82),
    "9876547": (9.19, 31.97, 14.55),
    "98965471": (98.28, 167.76, 377.35)
}

speedups = {}

# calculate speedups
for key, (naive, approach1, approach2) in execution_times.items():
    if naive != 0.0:
        speedup_approach1 = naive / approach1
        speedup_approach2 = naive / approach2

        # store the speedups
        speedups[key] = (speedup_approach1, speedup_approach2)

# dataframe for speedups
speedup_df = pd.DataFrame(speedups, index=["Speedup Approach 1", "Speedup Approach 2"]).T

display_table(name="Speedups Table", dataframe=speedup_df)

Speedup Approach 1  Speedup Approach 2
123456              0.750000            0.363636
987654              0.669118            0.322695
9876547             0.287457            0.631615
98965471            0.585837            0.260448

## Cálculo de eficiencia

In [5]:
# format: (naive time, approach 1 time, approach 2 time)
execution_times_with_processes = {
    "np 2": (0.13, 0.15, 0.15, 2),
    "np 3": (0.12, 0.18, 0.40, 3),
    "np 4": (0.11, 0.17, 0.34, 4),
}

efficiencies = {}

# calculate the speedups and then the efficiency
for key, (naive, approach1, approach2, processes) in execution_times_with_processes.items():
    if naive != 0.0:
        speedup_approach1 = naive / approach1
        speedup_approach2 = naive / approach2
        efficiency_approach1 = speedup_approach1 / processes
        efficiency_approach2 = speedup_approach2 / processes
        efficiencies[key] = (efficiency_approach1, efficiency_approach2)

# dataframe for efficiencies
efficiency_df = pd.DataFrame(efficiencies, index=["Efficiency Approach 1", "Efficiency Approach 2"]).T

display_table(name="Efficiencies Table", dataframe=efficiency_df)

Efficiency Approach 1  Efficiency Approach 2
np 2               0.433333               0.433333
np 3               0.222222               0.100000
np 4               0.161765               0.080882

## Análisis de resultados
Los resultados obtenidos muestran que el speedup logrado no es lineal, lo que indica que los recursos de paralelización no están siendo completamente aprovechados. Por ejemplo, con 2 procesos, el speedup en el enfoque 1 fue de 0.75, mientras que en el enfoque 2 fue de 0.36, lo que está muy por debajo del ideal. Esto sugiere que factores como la sobrecarga de comunicación y la carga no equilibrada entre procesos están afectando el rendimiento.

La eficiencia, que mide el uso efectivo de los procesadores, disminuye a medida que se incrementa el número de procesos. En la configuración de 4 procesos, la eficiencia cayó a 0.16 y 0.08 para los enfoques 1 y 2, respectivamente. Esto es consistente con la Ley de Amdahl, que establece que una porción secuencial del programa limita el rendimiento conforme se agregan más procesos.

En el contexto del proyecto de descifrado de claves con MPI, este comportamiento puede explicarse por la naturaleza de la búsqueda por fuerza bruta. Si la clave se encuentra en una sección del espacio de búsqueda asignada a un proceso específico, los demás procesos no contribuyen de manera significativa, afectando el speedup global.